### ISSUE TITLE
"Computation Time Analysis for TimeDistributed Layer – Issue #21179"

### ISSUE THREAD

**Issue by gorden-del** [keras-team/keras#21208]
Inference using TimeDistributed(Dense(1, activation="sigmoid")) in Keras 3 is significantly slower than in Keras 2 (tf.keras). Profiling indicates that the TimeDistributed layer is the primary bottleneck.

Model Architecture:
Conv1D → LSTM → Conv1DTranspose → TimeDistributed(Dense(1))
Input Shape: (None, None, 1)

What I've Tried:

run_eagerly=False

Wrapping in @tf.function

Switching between TensorFlow, JAX, and PyTorch backends

None of these attempts led to meaningful performance improvements.

Question:
Is there a way to resolve the TimeDistributed slowdown in Keras 3 without altering the model architecture?
(Current workaround: reverting to tf.keras.)

Environment:

Keras Version: 3.9.0

**Comment by dhantule** [comment]
Hi @gorden-del, thanks for reporting this. 

Could you provide some code to reproduce ?

**Comment by gorden-del** [comment]
Problem Description
When running inference on a sequence model (Conv1D → LSTM → Conv1DTranspose → TimeDistributed(Dense(1))), Keras 3 exhibits significantly slower performance compared to Keras 2 (tf.keras). Profiling reveals that the TimeDistributed layer is the primary bottleneck.

Key Observations
Input Shape: (None, None, 1) (variable-length sequences)

Backends Tested: TensorFlow, JAX, PyTorch

Attempted Optimizations:

run_eagerly=False

Wrapping inference in @tf.function

Switching between Keras backends

No meaningful improvement observed

Current Workaround: Reverting to tf.keras

Root Cause
Keras 3 introduces a more generalized implementation of TimeDistributed, which may introduce overhead due to:

Dynamic shape handling (slower in some backends)

Cross-backend compatibility checks

Potential inefficiencies in the JIT compilation (for JAX/PyTorch backends)

**Proposed Solutions**
1. Replace TimeDistributed with a Reshape + Dense Workaround
Since TimeDistributed(Dense(1)) applies the same Dense layer across time steps, we can manually flatten and reshape `instead:

`import keras  
from keras import layers  

def build_model():
    inputs = layers.Input(shape=(None, 1))  
    x = layers.Conv1D(64, 3, activation='relu')(inputs)  
    x = layers.LSTM(128, return_sequences=True)(x)  
    x = layers.Conv1DTranspose(64, 3, activation='relu')(x)  
    
    # Replace TimeDistributed(Dense(1)) with reshape + Dense  
    x = layers.Reshape((-1, 1))(x)  # Flatten time steps  
    x = layers.Dense(1, activation='sigmoid')(x)  
    x = layers.Reshape((-1, 1))(x)  # Restore original shape  
    
    return keras.Model(inputs=inputs, outputs=x)`

**Pros:**
Avoids TimeDistributed overhead
Works across all backends

**Cons:**
Requires matching shapes carefully

**2. Force TensorFlow Backend with Optimized Kernel (if using TF)**
If using TensorFlow backend, ensure it uses the optimized implementation:

`import os
os.environ["KERAS_BACKEND"] = "tensorflow"  # Force TF backend
import keras

# Build model as usual
model = build_model()  
model.compile(run_eagerly=False)  # Disable eager execution`

**Note: This may not fully resolve the issue but could improve performance.**

**3. Use jit_compile=True (For JAX/PyTorch Backends)**
If using JAX/PyTorch, forcing JIT compilation might help:

`model.compile(jit_compile=True)  # For JAX/PyTorch backends`

**4. Downgrade to Keras 2 (Last Resort)**
If performance is critical and other fixes don’t help:

`pip install "keras<3.0"  # Revert to tf.keras`

**Benchmarking & Validation**
Test inference speed before/after changes:

`import time
import numpy as np

# Generate test input
test_input = np.random.rand(1, 1000, 1)  # Batch of 1, 1000 timesteps, 1 feature

# Benchmark
start = time.time()
model.predict(test_input)
print(f"Inference time: {time.time() - start:.4f}s")`

**Conclusion**

**Best Short-Term Fix: Replace TimeDistributed with Reshape + Dense (Solution 1).**
**If Using TF Backend: Ensure run_eagerly=False and proper backend selection.**
**If Using JAX/PyTorch: Try jit_compile=True.**
**Final Fallback: Revert to tf.keras if necessary.**

Would you like additional profiling (e.g., keras.utils.trace_model) to identify exact bottlenecks?

**Comment by janwodnicki** [comment]
I recently migrated to Keras 3 and, like @gorden-del, identified the TimeDistributed layer as the main performance bottleneck when using TensorFlow 2.18.

The [first implementation](https://github.com/keras-team/keras/blob/6d06085d435e10846fe88c85053b36ae6b992462/keras/src/layers/rnn/time_distributed.py#L122) in `TimeDistributed.call` uses a Python for loop, which seems to cause TensorFlow to retrace the inner layer once per timestep.

When I comment out the first implementation and switch to the [second implementation](https://github.com/keras-team/keras/blob/6d06085d435e10846fe88c85053b36ae6b992462/keras/src/layers/rnn/time_distributed.py#L130) that uses `backend.vectorized_map`, my model compiles in a few seconds instead of over an hour.

@fchollet
Looking at the commit history, on [May 18, 2023](https://github.com/keras-team/keras/commit/5b1c0ed4440b8a5c5923c696a8abad4760a2bfc2) a `backend.vectorized_map`-only implementation was added, and the next day ([May 19, 2023](https://github.com/kerasteam/keras/commit/4dd75e0f2754dfa538d278c94a264d4b5d581033)) the Python for loop path was introduced.

I’m curious if the Python loop for static time dimensions was added for a specific backend or execution mode. It seems to introduce major overhead on TensorFlow, so it’d be great to understand the reasoning - maybe there’s a good opportunity here for an optimization or conditional path.

Thanks for your time and work on Keras!
